Glade

GuillaumeGomez opened the pull request:
Translate some examples to Glade.

Once this merged, the original examples (the non-Glade code) will be moved into its own repository (to keep tests on "normal" code).

Part of #92

gkoz commented on src/gtktest.rs:
```diff
@@ -3,295 +3,213 @@
+
+    /// Expands to its argument if GTK+ 3.10 support is configured and to `()` otherwise
+    #[cfg(not(feature = "gtk_3_10"))]
+    macro_rules! with_gtk_3_10 {
```
Isn't this macro redundant now?

GuillaumeGomez replied:
Worst: it was unused. Strange there wasn't a warning about it.

gkoz commented on src/gtktest.rs:
```diff
@@ -3,295 +3,213 @@
+        let builder = Builder::new_from_string(glade_src);
+
+        let window: gtk::Window = builder.get_object("window").unwrap();
+        let button: gtk::Button = builder.get_object("button").unwrap();
```
Better import `Window`, `Button` and so on. Also the code would probably be easier to understand if calls to `connect` were closer to `get_object`. This `button` is used like 30 lines from where it's created.

gkoz commented on src/gtktest.rs:
```diff
@@ -3,295 +3,213 @@
+            }
+        });
+
+        let entry_clone = entry.clone();
```
Looks like the `clone!` macro is needed here too.

gkoz commented on src/sync_widgets.rs:
```diff
@@ -4,46 +4,53 @@
-    if gtk::init().is_err() {
-        println!("Failed to initialize GTK.");
-        return;
+#[cfg(feature = "gtk_3_10")]
```
If the only reason for this is `Builder::new_from_string` we should rather bind the `add_from_string` method, which does not require 3.10.

gkoz commented on src/sync_widgets.rs:
```diff
@@ -4,46 +4,53 @@
+
+        let window: gtk::Window = builder.get_object("window").unwrap();
+        let spin_button: gtk::SpinButton = builder.get_object("spin_button").unwrap();
+        spin_button.set_range(0.0, 130.0);
```
This should be possible to set from Glade.

GuillaumeGomez replied:
Certainly, but I can't find where. :-/

GuillaumeGomez commented:
@gkoz: I think I addressed all your comments.

gkoz commented on src/sync_widgets.glade:
```diff
@@ -0,0 +1,43 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!-- Generated with glade 3.16.1 -->
+<interface>
+  <requires lib="gtk+" version="3.10"/>
```
This requirement doesn't apply now?

GuillaumeGomez replied:
Arf!

gkoz commented:
The sync_widgets examples also doesn't need gtk 3.10 I assume.

Can you set up adjustments in Glade instead of in the code?

EPashkin commented on src/gtktest.rs:
```diff
@@ -3,295 +3,189 @@
+        }
+        println!("Major: {}, Minor: {}", gtk::get_major_version(), gtk::get_minor_version());
+        let glade_src = include_str!("gtktest.glade");
+        let builder = Builder::new_from_string(glade_src);
```
As @gkoz says this not work in 3.10, while `add_from_string` works.
IMHO if no other controls need 3.10, we can
```
let builder = if cfg!(feature = "gtk_3_10")  {
 Builder::new_from_string(glade_src)
} else {
let builder = Builder::new();
builder.add_from_string(glade_src)
}
```
and remove other switches

GuillaumeGomez replied:
In this example, a few other things need 3.10 gtk version, so just changing this wouldn't be enough.

gkoz replied:
`cfg!` can't prevent build errors. That's the reason the `with_gtk_3_10` macro exists. Anyway it's not like the add methods are deprecated.

EPashkin replied:
> In this example, a few other things need 3.10 gtk version, so just changing this wouldn't be enough.

Strange, `gtktest` seems work after using `builder.add_from_string`.
Bad that `if cfg!` not work

EPashkin replied:
`text_viewer` also works, despite `<requires lib="gtk+" version="3.10"/>`
May be it checked only in runtime with real gtk version.

EPashkin replied:
.. and we can't just change building and assume that everything is fine

GuillaumeGomez replied:
So what do you want me to do?

EPashkin replied:
Sorry, was wrong, all normal here.

EPashkin commented on src/gtktest.rs:
```diff
@@ -3,295 +3,189 @@
+        }));
+
+        let button_about: Button = builder.get_object("button_about").unwrap();
+        // use a plain function instead of a closure
```
Its TODO?

GuillaumeGomez replied:
No idea, it was like that so I kept it.

gkoz replied:
This example sort of shows that you can also use functions as signal handlers.

EPashkin replied:
But it don't use function as signal handler, compare it with `drawing_area.connect_draw(draw_fn);` from `cairotest.rs`

gkoz replied:
You're right, then there's no point having the function. We don't really need those `println`s, they come from the time when I wasn't confident in array translations.

EPashkin replied:
My problem is that comment saying "I'm use function instead closure"  but still use closure, so better remove comment.
I don't see anything wrong with debug prints in example.

EPashkin commented on src/gtktest.rs:
```diff
@@ -3,295 +3,189 @@
+        }));*/ //TODO: Why does this not work?
+
+        let scale: Scale = builder.get_object("scale").unwrap();
+        scale.set_range(0., 100.);
```
To glade?

GuillaumeGomez replied:
Like I already said, I didn't find how to do it in glade directly.

gkoz replied:
It's not a big deal, none of us know GTK very well. It's puzzling that you'd go on not knowing and offload the googling to someone else.

EPashkin replied:
May be this help, change with `GtkAdjustment`?
http://stackoverflow.com/questions/30407433/gtkscale-draws-correctly-in-glade-not-in-the-final-app
Not checked.

GuillaumeGomez replied:
> It's not a big deal, none of us know GTK very well. It's puzzling that you'd go on not knowing and offload the googling to someone else.

That's quite offending that you might I just let things like this in hope someone will do it for me. I let it like this because I wasn't able to find the way to do it.

gkoz replied:
That's exactly what you're doing by claiming to have addressed all comments. Someone has to spend effort to find out if that is true and then solve the issue because clearly you're done with it.

GuillaumeGomez replied:
I guess you're right. Sorry, I finally found out.

EPashkin commented on src/gtktest.rs:
```diff
@@ -3,295 +3,189 @@
+        let spinner: Spinner = builder.get_object("spinner").unwrap();
+        spinner.start();
+
+        /*info_bar.connect(signals::Response::new(|response_id| {
```
Remove this comment?
In not worked as `window.show_all();` called after connect,
same effect can be done by glade if really needed. (not sure about it)

GuillaumeGomez commented:
Ok, so I updated. I removed `set_range` calls from code to add them to glade directly (damn adjustment thing). I removed unneeded comments. Do you see anything else?

EPashkin commented on src/gtktest.glade:
```diff
@@ -0,0 +1,552 @@
+<!-- Generated with glade 3.16.1 -->
+<interface>
+  <requires lib="gtk+" version="3.10"/>
+  <object class="GtkAdjustment" id="adjustment1">
```
Not better 'scale_adjustment' ?

EPashkin replied:
I talked about id, other ids for GtkAdjustment also not good.

GuillaumeGomez replied:
Since we don't use them, I didn't give them a "correct" name. I change it.

EPashkin commented on src/gtktest.rs:
```diff
@@ -3,295 +3,182 @@
+        let builder = Builder::new_from_string(glade_src);
+
+        let window: Window = builder.get_object("window").unwrap();
+        let entry: Entry = builder.get_object("entry").unwrap();
```
Can you move it down to use place?

EPashkin commented on src/gtktest.rs:
```diff
@@ -3,295 +3,182 @@
+        }));
+
+        let button_about: Button = builder.get_object("button_about").unwrap();
+        button_about.connect_clicked(move |x| {
```
First click was good, second show many asserts, empty window and crash on mouse over.

GuillaumeGomez replied:
It comes from the fact that Glade creates only one instance of it, so we need to show/hide instead of destroy. I added a comment in the source code to explain it.

GuillaumeGomez commented:
Updated.

EPashkin commented on src/gtktest.rs:
```diff
@@ -3,295 +3,184 @@
+        let glade_src = include_str!("gtktest.glade");
+        let builder = Builder::new_from_string(glade_src);
+
+        let window: Window = builder.get_object("window").unwrap();
```
Down to `let button` ?

EPashkin commented on src/sync_widgets.rs:
```diff
@@ -4,37 +4,28 @@
-    hbox.add(&slider);
-
-    window.add(&hbox);
     window.show_all();
```
Not sure if it wrong but it always near `gtk::main();` after all connects

GuillaumeGomez replied:
It's just to make life easier by avoid useless `clone` calls.

EPashkin replied:
I just moved it down and has no problem. `window` used only in `window.connect_delete_event` so also `get_object` can move there.

GuillaumeGomez replied:
Indeed, I moved it. Well saw!

EPashkin commented:
:+1: after fixing last commit name or split it to olds. Thanks.

GuillaumeGomez commented:
Forgot to squash :open_mouth:

EPashkin commented:
@GuillaumeGomez by way, you can use comment names like `fixup! Translate sync_widgets example to glade` and do rebase after squashing (if rebase.autoSquash = true set)

GuillaumeGomez commented:
Oh I didn't know. Thanks for the tip!

EPashkin commented:
* not `after squashing` but `after PR approved`

EPashkin commented:
IMHO this also helps to see what fixed

EPashkin commented:
Now last commit contains both `gtktest` and `sync_widgets` :(

GuillaumeGomez commented:
Ok, i rewrite the commits as well. :)

GuillaumeGomez commented:
@EPashkin: I rewrote commits. Better?

EPashkin commented on src/gtktest.glade:
```diff
@@ -0,0 +1,552 @@
+                        <property name="can_focus">True</property>
+                        <property name="receives_default">True</property>
+                        <property name="relief">none</property>
+                        <property name="uri">www.rust-lang.org</property>
```
https://www.rust-lang.org/

EPashkin replied:
`www.rust-lang.org` was it label

GuillaumeGomez replied:
You want me to change the label by the way?

EPashkin replied:
Yes `button` not good.
IMHO label better without `https://` as tooltip shows full address.

GuillaumeGomez replied:
Yes sir!

EPashkin commented:
Thanks

gkoz commented:
Looks good, except text_viewer also doesn't seem to need 3.10.

GuillaumeGomez commented:
@gkoz: Fixed.

gkoz commented:
Thanks for researching Glade! We can now think what can be done about ergonomics of this approach.
